crates/types/src/net/mod.rs

```diff
@@ -51,22 +52,57 @@ pub enum AdvertisedAddress {
     /// Unix domain socket
     #[display("unix:{}", _0.display())]
     Uds(PathBuf),
-    /// Hostname or host:port pair, or any unrecognizable string.
+    /// Hostname or host:port pair
     #[display("{}", _0)]
     Http(Uri),
 }
 
 impl FromStr for AdvertisedAddress {
-    type Err = http::uri::InvalidUri;
+    type Err = Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        if let Some(stripped_address) = s.strip_prefix("unix:") {
-            Ok(AdvertisedAddress::Uds(
-                stripped_address.parse().expect("infallible"),
-            ))
-        } else {
-            // try to parse as a URI
-            Ok(AdvertisedAddress::Http(s.parse()?))
+        match s.trim() {
+            "" => Err(anyhow::anyhow!("Advertised address cannot be empty")),
+            address if address.starts_with("unix:") => {
+                parse_uds(&address[5..]).map(AdvertisedAddress::Uds)
+            }
+            address => parse_http(address),
+        }
+    }
+}
+
+fn parse_uds(s: &str) -> Result<PathBuf, Error> {
+    s.parse::<PathBuf>()
+        .with_context(|| format!("Failed to parse Unix domain socket path: '{}'", s))
+}
+
+fn parse_http(s: &str) -> Result<AdvertisedAddress, Error> {
+    let uri = s
+        .parse::<Uri>()
+        .with_context(|| format!("Invalid URI format: '{}'", s))?;
+
+    match uri.scheme_str() {
+        Some("http") | Some("https") => Ok(AdvertisedAddress::Http(uri)),
+        Some(other) => Err(anyhow::anyhow!("Unsupported URI scheme '{}'", other)),
+        None => Err(anyhow::anyhow!("Missing URI scheme in: '{}'", s)),
+    }
+}
+
+impl AdvertisedAddress {
+    /// Derives a `BindAddress` based on the advertised address.
+    pub fn derive_bind_address(&self) -> Option<BindAddress> {
```

Comment:
Nice!

One question though; should we not _always_ be able to derive a bind address, given a valid advertised address? If we have successfully constructed a `AdvertisedAddress` I would hope that to be the case!

Reply:
only in one case we can't derive the bind address, when advertised address is `http://example.com` and hasn't port specified, as I saw a comment on the advertised address ` Http(Uri)` /// Hostname or host:port pair

in case of hostname only, should we fallback to default port? 
or should we refactor to prevent parsing advertised address without port specified?

Reply:
Let's default this to `5122` and update the doc string to say that if advertised address doesn't have a port, we'll use that? :-)

This way we don't need to call `unwrap()` on a potential `None` option in network_server/service.rs.

Reply:
Done 🔢